swarm/storage/resource.go

```diff
@@ -0,0 +1,528 @@
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+/***
+*
+* Resource updates is a data update scheme built on swarm chunks
+* with chunk keys following a predictable, versionable pattern.
+*
+* The intended use case is to update data locations of certain
+* hashes
+*
+* Updates are defined to be periodic in nature, where periods are
+* expressed in terms of number of blocks.
+*
+* The root entry of a resource update is tied to a unique identifier,
+* typically - but not necessarily - an ens name. It also contains the
+* block number when the resource update was first registered, and
+* the block frequency with which the resource will be updated.
+* Thus, a resource update for identifier "foo.bar" starting at block 4200
+* with frequency 42 will have updates on block 4242, 4284, 4326 and so on.
+*
+* the identifier is supplied as a string, but will be IDNA converted and
+* passed through the ENS namehash function. Pure ascii identifiers without
+* periods will thus merely be hashed.
+*
+* The data format of the root entry is the ENS namehash as key and
+* blocknumber and frequency in little-endian uint64 representation as values,
+* concatenated in that order, for a total of 16 bytes.
+*
+* Note that the root entry is not required for the resource update scheme to
+* work. A normal chunk of the blocknumber/frequency data can also be created,
+* and pointed to by an actual ENS entry instead.
+*
+* Actual data updates are also made in the form of swarm chunks. The keys
+* of the updates are the hash of a concatenation of properties as follows:
+*
+* sha256(namehash|blocknumber|version)
+*
+* The blocknumber here is the next block period after the current block
+* calculated from the start block and frequency of the resource update.
+* Using our previous example, this means that an update made at block 4285,
+* and even 4284, will have 4326 as the block number.
+*
+* If more than one update is made to the same block number, incremental
+* version numbers are used successively.
+*
+* A lookup agent need only know the identifier name
+*
+* the data itself is prefixed with a signed hash of the data. The sigining key
+* can be used to verify the authenticity of the update, for example by looking
+* up the ownership of the namehash in ENS and comparing to the address derived
+* from it
+*
+* NOTE: the following is yet to be implemented
+* The resource update chunks will be stored in the swarm, but receive special
+* treatment as their keys do not validate as hashes of their data. They are also
+* stored using a separate store, and forwarding/syncing protocols carry per-chunk
+* flags to tell whether the chunk can be validated or not; if not it is to be
+* treated as a resource update chunk.
+*
+* TODO: signature validation
+ */
+
+const (
+	SIGNATURE_LENGTH = 65
+)
+
+// Encapsulates an actual resource update. When synced it contains the most recent
```

Comment:
I think we should use godoc convention for all source code, as it is used throughout the go-ethereum codebase - https://blog.golang.org/godoc-documenting-go-code , which can then be indexed nicely in godoc.org . This applies for all the comments in the PR.